### PR TITLE
[Widget enhancement] 04 - Subscribe to cow events (from iframe)

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

Part of the series https://github.com/cowprotocol/cowswap/pull/3812

This PR creates a `subscribeToCoWEvents` which will evolve later into a class `IframeCowEventEmitter`! (check out last PR for that)

Also continue shaping the types of the events. I'm not done with this yet, future PRs will address issues with current types.

### CODE REVIEW

File: libs/events/src/EventEmitter.ts [hunk 1]
```diff
@@ -1,34 +1,47 @@
-import { EventNames, EventPayloads } from './types'
+import { CowEvents, EventPayloads as CowEventPayloads } from './types'
```
Comment (alfetopito) [inline libs/events/src/EventEmitter.ts:1]: Did you mean to rename the source `EventPayloads` type as well?

File: libs/events/src/types/events.ts [hunk 2]
```diff
@@ -1,13 +1,18 @@
...
+// export type CowEvents = 'updateWidgetConfig' | 'onToastMessage' | 'onPostedOrder' | 'onExecutedOrder'
+export enum CowEvents {
+  ON_TOAST_MESSAGE = 'ON_TOAST_MESSAGE',
+  ON_POSTED_ORDER = 'ON_POSTED_ORDER',
+  ON_EXECUTED_ORDER = 'ON_EXECUTED_ORDER',
+  ON_REJECTED_ORDER = 'ON_REJECTED_ORDER',
+}
 
 // Define types for event payloads
 export interface EventPayloads {
```
Comment (alfetopito) [inline libs/events/src/types/events.ts:13]: What about a Record<CowEvents, PayloadBaseType> to make sure no enum property is forgotten?

File: libs/widget-lib/src/cowSwapWidget.ts [hunk 3]
```diff
@@ -137,3 +145,29 @@ function applyDynamicHeight(iframe: HTMLIFrameElement, defaultHeight = DEFAULT_H
...
+ * @param events - subscription to different events
+ */
+function subscribeToCoWEvents(events?: CowEventSubscriptions): void {
+  if (!events) {
+    return
+  }
+
+  const eventEmitter = new CowEventEmitterImpl()
+
+  window.addEventListener('message', (event) => {
```
Comment (alfetopito) [inline libs/widget-lib/src/cowSwapWidget.ts:161]: Is there a type for this `event`?

File: libs/events/src/EventEmitter.ts (hunk 1 above)
Comment (anxolin) [reply libs/events/src/EventEmitter.ts:1]: Yes, i saw this later, is not present in the most recent PRs

File: libs/widget-lib/src/cowSwapWidget.ts (hunk 3 above)
Comment (anxolin) [reply libs/widget-lib/src/cowSwapWidget.ts:161]: No, it's kind of a standard https://developer.mozilla.org/en-US/docs/Web/API/Window/message_event

File: libs/events/src/types/events.ts (hunk 2 above)
Comment (anxolin) [reply libs/events/src/types/events.ts:13]: Very good point! I will do so

Comment (anxolin) [reply libs/events/src/types/events.ts:13]: mm, I don't know how to do this one. `Record<CowEvents, PayloadBaseType>` don't work as each payload is independent, you need to define it on a per-case basis. 

Shouldn't be too bad though. When you add an event, you will notice you need to add it, and if you forget, you get already now a typescript error:

I've tried:
image

anyways, if you have an idea how to improve, let me know

Comment (alfetopito) [reply libs/events/src/types/events.ts:13]: I don't, but maybe ChatGPT does =P
Or even better, @shoom3301 might!

Comment (shoom3301) [reply libs/events/src/types/events.ts:13]: I actually don't know, but I asked one nerd guy about it and he gave me this:
https://gist.github.com/lebedev/60cd207a5985df62c943883aed27e499

I'm not sure does it worth it or not :)

Comment (alfetopito) [reply libs/events/src/types/events.ts:13]: Ok, I'm fine with that then

Comment (anxolin) [reply libs/events/src/types/events.ts:13]: Ohh, nice, thanks both!!!

Comment (anxolin) [reply libs/events/src/types/events.ts:13]: 980bdf7a5

Comment (anxolin) [reply libs/events/src/types/events.ts:13]: 980bdf7a5

File: libs/widget-lib/src/types.ts [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import type { SupportedChainId } from '@cowprotocol/cow-sdk'
 export type { SupportedChainId } from '@cowprotocol/cow-sdk'
+import type { CowEventEmitter } from '@cowprotocol/events'
```
Comment (shoom3301) [inline libs/widget-lib/src/types.ts:3]: Lint error

**Review by shoom3301 — APPROVED** [review]

File: libs/widget-lib/src/types.ts (hunk 4 above)
Comment (anxolin) [reply libs/widget-lib/src/types.ts:3]: I think in current PRs this was deleted

**Pull request merged**